### PR TITLE
feat: add new useMultichainSelector

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This PR introduces `useMultichainSelector` to pass an additional account arg when using certain multichain selectors. 

## **Related issues**

Fixes https://github.com/MetaMask/accounts-planning/issues/495

## **Manual testing steps**

This PR only adds selectors and they are not currently used on the extension. It will setup the following PR https://github.com/MetaMask/metamask-extension/pull/25437

## **Screenshots/Recordings**

No UI/UX changes

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/selectors/accounts.test.ts [hunk 1]
```diff
@@ -47,4 +51,36 @@ describe('Accounts Selectors', () => {
...
+      'returns $isBtc if the account is: $type',
+      ({ id, isBtc }: { id: string; isBtc: boolean }) => {
+        const state = MOCK_STATE;
+
+        state.metamask.internalAccounts.selectedAccount = id;
+        expect(isSelectedInternalAccountBtc(state)).toBe(isBtc);
+      },
+    );
+
+    it('returns false if none account is selected', () => {
```
Comment (k-g-j) [inline ui/selectors/accounts.test.ts:79]:
```suggestion
    it('returns false if no account is selected', () => {
```

File: ui/selectors/multichain.ts [hunk 2]
```diff
@@ -108,11 +123,14 @@ export function getMultichainNetwork(
 // a popup (for ethereum related stuffs) is being shown (and uses this function), then the native
 // currency will be BTC..
 
-export function getMultichainIsEvm(state: MultichainState) {
+export function getMultichainIsEvm(
```
Comment (k-g-j) [inline ui/selectors/multichain.ts:126]: Do we have any tests for these new selectors?

File: ui/selectors/accounts.ts [hunk 3]
```diff
@@ -12,3 +12,10 @@ export function isSelectedInternalAccountEth(state: AccountsState) {
 
   return Boolean(account && (account.type === Eoa || account.type === Erc4337));
 }
+
+export function isSelectedInternalAccountBtc(state: AccountsState) {
```
Comment (ccharly) [inline ui/selectors/accounts.ts:16]: Should not be part of this PR

Comment (montelaidev) [reply ui/selectors/accounts.ts]: Removed in daa5e47fb99450b9abc38f37f67a94349a733a75

[the remark this answers is not in the document]
Comment (montelaidev) [reply ui/selectors/accounts.ts]: Removed in daa5e47fb99450b9abc38f37f67a94349a733a75

File: ui/selectors/accounts.test.ts [hunk 4]
```diff
@@ -40,11 +44,43 @@ describe('Accounts Selectors', () => {
...
-    it('returns false if none account is selected', () => {
+    it('returns false if no account is selected', () => {
       const state = MOCK_STATE;
 
       state.metamask.internalAccounts.selectedAccount = '';
       expect(isSelectedInternalAccountEth(MOCK_STATE)).toBe(false);
     });
   });
+
+  describe('isSelectedInternalAccountBtc', () => {
```
[the remark this answers is not in the document]
Comment (montelaidev) [reply ui/selectors/accounts.test.ts]: Removed in daa5e47fb99450b9abc38f37f67a94349a733a75

File: ui/selectors/multichain.ts [hunk 5]
```diff
@@ -30,7 +30,20 @@ import {
 
 export type MultichainState = AccountsState & {
   metamask: {
-    // TODO: Use states from new {Rates,Balances,Chain}Controller
+    balances: {
```
Comment (ccharly) [inline ui/selectors/multichain.ts:33]: I wonder if we could use the same pattern than for `AccountsState` and use something like:

```ts
export type MultichainState = {
  metamask: BalancesControllerState & RatesControllerState & AccountsControllerState,
};
```

WDYT?

File: ui/selectors/multichain.ts [hunk 6]
```diff
@@ -108,15 +122,18 @@ export function getMultichainNetwork(
...
   const isOnboarded = getCompletedOnboarding(state);
   // Selected account is not available during onboarding (this is used in
   // the AppHeader)
-  const selectedAccount = getMaybeSelectedInternalAccount(state);
+  const selectedAccount = account ?? getMaybeSelectedInternalAccount(state);
 
   // There are no selected account during onboarding. we default to the original EVM behavior.
   return (
-    !isOnboarded || !selectedAccount || isEvmAccountType(selectedAccount.type)
+    !isOnboarded ?? !selectedAccount ?? isEvmAccountType(selectedAccount.type)
```
Comment (ccharly) [inline ui/selectors/multichain.ts:136]: This is not correct, we want "logical OR" here

**Review by ccharly — CHANGES_REQUESTED** [review]

File: ui/selectors/multichain.ts (hunk 5 above)
Comment (montelaidev) [reply ui/selectors/multichain.ts:33]: Yea, i think in the long term we should fix the `CombinedBackgroundAndReduxState` to have it be properly typed. We can do this in another PR

File: ui/selectors/multichain.ts (hunk 6 above)
Comment (montelaidev) [reply ui/selectors/multichain.ts:136]: Good catch

File: ui/selectors/multichain.ts [hunk 7]
```diff
@@ -28,12 +28,32 @@ import {
   getShouldShowFiat,
 } from '.';
 
-export type MultichainState = AccountsState & {
+export type RatesState = {
   metamask: {
-    // TODO: Use states from new {Rates,Balances,Chain}Controller
+    rates: {
```
Comment (ccharly) [inline ui/selectors/multichain.ts:33]: What about using `RatesControllerState`?

File: ui/selectors/multichain.ts [hunk 8]
```diff
@@ -28,12 +28,32 @@ import {
...
+        conversionDate: number;
+        conversionRate: string;
+      };
+    };
+  };
+};
+
+export type BalancesState = {
+  metamask: {
+    balances: {
```
Comment (ccharly) [inline ui/selectors/multichain.ts:44]: What about using `BalancesControllerState`?

**Review by ccharly — DISMISSED** [review]

**Review by ccharly — APPROVED** [review]

**Review by owencraston — APPROVED** [review]
